src/emulate.c

```diff
@@ -103,7 +95,7 @@ static void trap_handler(riscv_t *rv);
  * identifier called tval, as both are handled by TRAP_HANDLER_IMPL.
  */
 #define TRAP_HANDLER_IMPL(type, code)                                         \
-    static void rv_trap_##type(riscv_t *rv, uint32_t tval)                    \
+    void rv_trap_##type(riscv_t *rv, uint32_t tval)                           \
```

Comment:
It is a pity that the helper functions can not be declared as `static`. Can you find a way for that?

Reply:
> It is a pity that the helper functions can not be declared as `static`. Can you find a way for that?

We could register the `trap_handler` with the I/O interface during RISC-V core initialization. This way, the dispatch table can remain hidden. When page fault occurs, `trap_handler` is called and then it calls `_trap_handler` which might be the helper trap helper function defined in emulate.c . But, the `_trap_handler` might needs to check the `scause` CSR to determine which `TRAP_HANDLER_IMPL` should be called.

Reply:
In detail,  `trap_handler` callback (in IO interface) —> `_trap_handler`(check scause CSR  and dispatch `TRAP_HANDLER_IMPL` depends on scause CSR, static function in emulate.c)—> `__trap_handler`(actual trap handler, the current `trap_handler`, static function in emulate.c)